### ISSUE TITLE
[Discussion] What do you want to see in decaffeinate?

### ISSUE THREAD

**Issue by alangpierce** [decaffeinate/decaffeinate#403]
This repository has been a bit quiet recently, so I wanted to start an informal discussion to see if anyone out there has any input about the project or its high-level direction.

I guess decaffeinate is sort of an interesting project since it's meant for a one-time conversion, and once you finish using it, there isn't much need to stay involved in the project anymore. So it's certainly possible that the project is less active because people have successfully moved off of CoffeeScript (and that would be great!), but if there are others out there with opinions or thoughts, it would be good to hear, and I want to make sure there's a place for that.

Some possible discussion questions:

* Have you used decaffeinate successfully? It would be great to hear any success stories, as well as any big issues you ran into. Moving a codebase from one programming language to another has a lot of challenges, both from an engineering standpoint and politically, so it would be good to hear people's experiences and any advice.
* Are you planning/hoping to use decaffeinate in the future? If so, how much code do you have to convert? Do you have a strategy in mind for how to convert the code?
* Are there any specific issues stopping you from using decaffeinate today?
* Are there any outstanding bugs that you think are particularly high-priority?

In my case, the codebase at my company has about 200,000 lines of CoffeeScript code that we eventually want to convert to JavaScript. There's probably no way to avoid a bunch of manual work, but I want to automate as much as possible, and decaffeinate is a core part of that. I've already converted about 3000 lines to JS, and I'm trying to make it as easy as possible to do larger and larger conversions. Currently, my focus is on fixing correctness issues in the generated JS, since I want my team to be confident that the conversion is safe. My longer-term goal is to be able to run decaffeinate on the whole codebase, then run all tests and have them all pass. My impression is that the project is already fairly stable, and it's ready to convert any particular file in my codebase (with some manual work), but there's a long tail of smaller cases that would be nice to fix.

Also, if anyone is interested in contributing, let me know and I can point you to some good starter bugs and explain how things work.

**Comment by madskonradsen** [comment]
I guess all projects go through quiet phases :) I have a project with around 1.500 CoffeeScript files, and is currently holding off with converting until it stabilizes a bit more, even though most of it could be fixed manually :) For me, it seems like the biggest issue atm. is the handling of parenthesis which get misplaced, but I guess it's a bit tricky to tackle all corner-cases.

**Comment by eventualbuddha** [comment]
I haven't contributed much recently for a few reasons:

* my personal life has been rather busy lately
* I've been sick (just a cold for the last week)
* with the addition of the `__guard__` helper to handle soaks, decaffeinate is now "good enough" to not crash on almost all files in the main codebase we're converting at Square

That third one is probably the biggest reason. There are a few things I want to do such as adding an option to prefer `const` (#63). And a lot of people converting files at Square have checked in code with the closing parenthesis on the next line after a closing curly brace:

```js
}
);
```

It's annoying to have to review for such cases. I've also found that it'd be nice to define custom rules for when to add implicit `return`s (#409).

**Comment by aristidesfl** [comment]
Would it be interesting to reuse this project to create a drop in replacement for https://github.com/babel/babylon so that we could have a coffeescript parser compatible with babel?

**Comment by alangpierce** [comment]
@aristidesfl I see what you're saying, but decaffeinate isn't really meant to be a transform you run as part of a build step (like you usually do with babel transforms). Compared with the CoffeeScript compiler, decaffeinate is much slower, crashes in some rare edge cases, and sometimes sacrifices 100% correctness, but the benefit is that it creates code that looks much nicer. It also never really generates a JavaScript AST; it uses the CoffeeScript AST to make in-place transformations on the code to turn it into JavaScript. Admittedly, I haven't worked with babel a lot, but it seems like it would be better to use the CoffeeScript compiler itself as a starting point for a babel-compatible CoffeeScript parser.

**Comment by hexsprite** [comment]
have been converting some files earlier today

nice to have:
- ES6 class support
- use ES6 `for X of ARR` to iterate through arrays
- optional use of underscore for certain operations like `_.contains(ARR, elem)`. I imagine this is a matter of style so having some kind of config would make sense

**Comment by alangpierce** [comment]
@hexsprite cool, thanks for the feedback. For reference, those issues are (roughly) #397, #323, and #363. Funny enough, I have all of those implemented and they should be shipping soon.

One particular question I have is with the `in` operator. What would you think about using `Array.prototype.includes` (from ES2016) instead of `_.contains`? If you need to support older browsers, you can polyfill it with [babel-polyfill](https://babeljs.io/docs/usage/polyfill/), more specifically the [stage 4 section](https://github.com/zloirock/core-js/blob/master/stage/4.js) of core-js.

My current plan is to change decaffeinate to output `includes` and to update the docs to mention that you need it, but if that's a pain I can hold off or make it a preference.

**Comment by madskonradsen** [comment]
ES7 and Array.prototype.includes are officially approved by the committee, and I guess that every serious JS project do already have some kind of pipeline for converting ES6/ES7 down to ES5 or whatever is needed. Wouldn't it be cleaner to have jscodeshift doing stuff like converting includes to underscore-fuunction rather than introducing too many options to decaffeinate :)

**Comment by hexsprite** [comment]
@madskonradsen oh, i forgot about `includes`. Nevermind that part then.

**Comment by alangpierce** [comment]
Here's an update on my progress and how I'm thinking about the project moving forward:

decaffeinate is now able to run on every file in my codebase without crashing, and the next milestone I'm working toward is getting all of the tests working after decaffeinate runs. Hopefully I can get that into a good state, and my plan after that is to focus on automating as much follow-up cleanup as possible. I also generally am trying to keep up with bugs as they are filed, even if they don't affect my codebase.

At my company, decaffeinate is ready for people to use in a self-serve way (although I've done all of the code reviews). I have a [bulk-decaffeinate](https://github.com/alangpierce/bulk-decaffeinate) config file set up in the repo and some internal documentation on how to use it, and out of the 16 other engineers at my company, 6 of them have used decaffeinate at least once. It's particularly useful when people know that they're soon going to be working in an area of code that's currently CoffeeScript; by running decaffeinate up-front, they can make sure new code is written in JS, and working with the code guards somewhat against possible problems introduced by decaffeinate. So far, we've collectively converted about 100 of the 1300 files originally in CoffeeScript, although most commits just converted one or two files at a time and there were a small number of large changes. While many of the steps have been automated, there is still quite a bit of manual work required for each file to clean up style issues to get lint to pass.

In trying to get my codebase's full test suite to pass, I ran into some challenges:

* The conversion from commonjs to JS modules introduces some bugs:
  * Imports and exports don't always match after the conversion. I was able to fix this with the fix-imports step in bulk-decaffeinate.
  * Unlike `require`, JS imports are evaluated before everything else in a file, so some environment setup code was running out of order, e.g. an import running before global state is modified rather than after. Since this is relatively rare, I could probably fix these cases manually, but it may be best for esnext to be more conservative here and skip any imports that appear after non-import code that may have side-effects.
  * Stubbing across module boundaries within tests doesn't always work after the conversion, particularly when switching to named imports/exports. In my codebase, that means that we need to change the tests from using sinon to using proxyquire, and the two systems are different enough that I probably can't automate the conversion. Probably there isn't a way for decaffeinate to help with this (although maybe a pref to avoid switching to named imports/exports would be nice). My current eventual goal is to see if I can come up with (or find) an alternative to proxyquire that lets you stub module fields in a more direct way.
* 11 files in my codebase use `this` before `super` in a class constructor, which is allowed in CS but not JS (it causes a babel error). I filed #591 to come up with some answer to this, but I suspect that the best we'll be able to do is document the issue and I can manually refactor those 11 files.

To work around both of these temporarily, I'm using the new `--keep-commonjs` option to skip the imports code altogether, and I'm skipping the 11 files that use `this` before `super`. With those changes, my tests can run without breaking catastrophically, and currently there are about 100 failures in about 5000 total tests. I've started looking into those and I think most of them should be relatively straightforward decaffeinate fixes (most likely much fewer than 100 distinct issues).

Generally, I think things are getting to a point where there are some unclear decisions about the tradeoff between code that looks nice and code that's completely correct. For example, a bunch of tests are failing because of differences between default params in CS vs JS (#125), and I've already accidentally introduced some bugs because of differences between arrays and array-like objects (#523).

The sentiment that I've generally seen in discussions both in decaffeinate issues and at my company is that correctness is way more important than nice-looking code. Still, ugly code tends to mean a lot of additional clean-up work. For example, at the moment we never leave `__guard__` calls in the JS code for soak operations; instead, we always rewrite the code by hand, and we may end up doing the same for default params and loops after #125 and #523 are addressed.

I think it's unclear whether decaffeinate should be viewed as a tool to mass-convert a codebase automatically or something that always needs sanity-checking and cleanup by a human (or something in between). It's nice to end up with clean JS code, but at the current (super-approximate) rate of 100 files every 4 months, we won't be off of CoffeeScript until 2021. To speed this up, I think the right approach may be to make decaffeinate more defensive, run it in larger batches, and care less about lint issues, but I think we'll need to experiment.

I'm thinking now that it's best to bias toward correctness by default, and expose "loose" options (naming inspired by babel loose mode) to make code cleaner-looking in specific cases. (Still, the focus on correctness shouldn't be excessive; we don't need to generate `void 0` instead of undefined.) For example, there might be `--loose-for-of`, `--loose-for-expressions`, `--loose-in`, `--loose-default-params`, `--loose-soaks`, etc. Generally, people might enable these loose modes when they are converting code in a more supervised way and they understand the implications. Making the behavior toggleable also makes it easier to explore which issues come up in real-world codebases, and perhaps some loose modes could be enabled by default if it's deemed safe enough.

So that's what I have in mind going forward. If anyone has any thoughts, ideas, or feedback, happy to discuss.

**Comment by sandstrom** [comment]
@eventualbuddha Just a short note to show my appreciation for this project.

We've now almost completed our coffee -> js conversion. We did it in chunks (models, helpers, library, etc) and `decaffeinate` has been a huge timesaver. Sure, there has still been some manual work (I think that's unavoidable) but the bulk has been automatic.

I'm very grateful!🥇 🔥  🎉

**Comment by hexsprite** [comment]
hey guys... is there a flag or how possible is it to customize the output so that  `@foo = 1` becomes `this.foo = 1` ?

I'm converting a Meteor project and this makes a lot more sense. I suppose if not I could run a post-conversion regex to fixup.

**Comment by alangpierce** [comment]
@hexsprite you can specify `--keep-commonjs` and it'll convert it to `this.foo = 1` (and skip all of the other import/export conversion). Does that work for you?

Note that top-level `this` is probably a bad idea in new code anyway (see #204). In the browser, it means `window`, and in node, it means `module.exports`. The conversion to `export let` was assuming the second interpretation. babel converts top-level `this` to `undefined` since it's not defined in the JS modules spec.

**Comment by hexsprite** [comment]
@alangpierce yes, but Meteor actually ends up internally wrapping each file such that top level `this` exists consistently and declares a global. So in my case it will work. this should work, thanks!

**Comment by alangpierce** [comment]
@hexsprite I see, makes sense. FYI I've been considering changing it so `--keep-commonjs` is the default behavior, and this seems like more motivation to do so.

**Comment by mreinstein** [comment]
@alangpierce my wishlist:

* ability to prepend this automatically to the beginning of each script file:

```javascript
'use strict'

```

* recursive directory traversal
* option to generate javascript with omitted semicolon line endings

**Comment by arodik** [comment]
Optional chaining support. This missing feature stops me from migrating my projects :(

**Comment by aminya** [comment]
> Optional chaining support. This missing feature prevents me from migrating my projects :(

Optional chaining is supported now. [Here's an example from the CoffeeScript web page.](https://decaffeinate-project.org/repl/#?useCS2=true&useJSModules=false&loose=false&optionalChaining=true&evaluate=false&stage=full&code=zip%20%3D%20lottery.drawWinner%3F().address%3F.zipcode%0A)
